File: Candy.cpp
class Solution {
public:

    int candy(vector<int> &ratings) {
        
        if (ratings.empty())
            return 0;
        if (ratings.size() == 1)
            return 1;
        
        int sum = 0;
        int cnt = 1;
        
        int i = 0;
        int j = 1;
        int len = ratings.size();
        
        bool isVisited = true;
        
        while(j < len) {
            if (ratings[i] < ratings[j]) {
                
                
                if (!isVisited)
                    cnt = 2;
                sum += cnt;
                isVisited = true;
                // sum += cnt;
                ++cnt;
            }
            else if (ratings[i] > ratings[j]){
                
                if (isVisited)
                    cnt = 1;
                sum += cnt;
                isVisited = false;
                // sum += cnt;
                ++cnt;
            }
            else {
                sum += cnt;
            }
            ++i;
            ++j;
        }
        
        // add last member of i
        if (ratings[i] > ratings[i-1] || ratings[i] == 1) {
            sum += cnt;
        }
        else
            sum += 1;
        return sum;
        
    }
    
};

File: jiuzhang/candy/candy.py
"""
 * 本代码由九章算法编辑提供。没有版权欢迎转发。
 * - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
 * - 现有的面试培训课程包括：九章算法班，系统设计班，九章强化班，Java入门与基础算法班
 * - 更多详情请见官方网站：http://www.jiuzhang.com/
"""

class Solution:
    # @param ratings, a list of integer
    # @return an integer
    def candy(self, ratings):
        candynum = [1 for i in range(len(ratings))]
        for i in range(1, len(ratings)):
            if ratings[i] > ratings[i-1]:
                candynum[i] = candynum[i-1] + 1
        for i in range(len(ratings)-2, -1, -1):
            if ratings[i+1] = candynum[i]:
                candynum[i] = candynum[i+1] + 1
        return sum(candynum)

